linera-execution/src/runtime.rs

```diff
@@ -1030,51 +1042,82 @@ impl<UserInstance> Clone for SyncRuntimeHandle<UserInstance> {
 }
 
 impl ContractSyncRuntime {
-    /// Main entry point to start executing a user action.
-    #[expect(clippy::too_many_arguments)]
-    pub(crate) fn run_action(
+    pub(crate) fn new(
         execution_state_sender: ExecutionStateSender,
-        application_id: UserApplicationId,
         chain_id: ChainId,
         local_time: Timestamp,
         refund_grant_to: Option<Account>,
         resource_controller: ResourceController,
-        action: UserAction,
+        action: &UserAction,
         txn_tracker: TransactionTracker,
-    ) -> Result<(ResourceController, TransactionTracker), ExecutionError> {
-        let executing_message = match &action {
-            UserAction::Message(context, _) => Some(context.into()),
-            _ => None,
-        };
-        let signer = action.signer();
-        let height = action.height();
-        let mut runtime = SyncRuntime(Some(ContractSyncRuntimeHandle::from(
+    ) -> Self {
+        SyncRuntime(Some(ContractSyncRuntimeHandle::from(
             SyncRuntimeInternal::new(
                 chain_id,
-                height,
+                action.height(),
                 local_time,
-                signer,
-                executing_message,
+                action.signer(),
+                if let UserAction::Message(context, _) = action {
+                    Some(context.into())
+                } else {
+                    None
+                },
                 execution_state_sender,
                 refund_grant_to,
                 resource_controller,
                 txn_tracker,
             ),
-        )));
+        )))
+    }
+
+    // TODO: deduplicate `Contract`/`Service` traits and types
+    pub(crate) fn preload_contract(
+        &self,
+        id: UserApplicationId,
+        code: UserContractCode,
+        description: UserApplicationDescription,
+    ) -> Result<(), ExecutionError> {
+        let this = self
+            .0
+            .as_ref()
+            .expect("contracts shouldn't be preloaded while the runtime is being dropped");
+        let runtime_handle = this.clone();
+        let mut this_guard = this.inner();
+
+        if let std::collections::hash_map::Entry::Vacant(e) =
+            this_guard.loaded_applications.entry(id)
+        {
+            e.insert(LoadedApplication::new(
```

Comment:
Nit: avoiding single letter bindings

```suggestion
        if let std::collections::hash_map::Entry::Vacant(entry) =
            this_guard.loaded_applications.entry(id)
        {
            entry.insert(LoadedApplication::new(
```